src/centralized.rs

```diff
@@ -0,0 +1,504 @@
+        utils::oracle_uniform(&digest, setup.n_p)
+    }
+
+    /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0 otherwise
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut data = vec!["Telescope-H2".as_bytes().to_vec()];
+        data.push(r.v.to_ne_bytes().to_vec());
+        data.push(r.t.to_ne_bytes().to_vec());
+        for s in &r.s_list {
+            data.push(s.clone().to_vec());
+        }
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(data);
+        utils::oracle_binomial(&digest, setup.q)
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &Vec<Vec<Element>>,
+        round: &Round,
+        limit: usize,
+    ) -> (usize, Option<Proof>) {
+        if round.s_list.len() == setup.u {
+            if Proof::h2(setup, round) {
+                let r = round.v;
+                let d = round.t;
+                let items = round.s_list.clone();
+                return (limit, Some(Proof { r, d, items }));
+            } else {
+                return (limit, None);
+            }
+        }
+        let mut dfs_limit = limit;
+        for &s in bins[round.h_usize].iter() {
+            if dfs_limit == setup.b {
+                return (dfs_limit, None);
+            }
+            let (l, proof_opt) = Self::dfs(setup, bins, &Round::update(round, s), dfs_limit + 1);
+            if proof_opt.is_some() {
+                return (l, proof_opt);
+            }
+            dfs_limit = l;
+        }
+        (dfs_limit, None)
+    }
+
+    /// Indexed proving algorithm, returns an empty proof if no suitable
+    /// candidate is found within the setup.b steps.
+    fn prove_index(setup: &Setup, set: &[Element], v: u32) -> (usize, Option<Proof>) {
+        let mut bins: Vec<Vec<Element>> = vec![];
+        for _ in 0..setup.n_p {
+            bins.push(vec![]);
+        }
+        for &s in set.iter() {
+            bins[Proof::h0(setup, v, s)].push(s);
+        }
+        let mut limit = 0;
+        for t in 0..setup.d {
+            if limit == setup.b {
+                return (0, None);
+            }
+            let round = Round::new(v, t + 1, setup.n_p);
+            let (l, res) = Proof::dfs(setup, &bins, &round, limit + 1);
+            if res.is_some() {
+                return (l, res);
+            }
+            limit = l;
+        }
+        (limit, None)
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Calls up to setup.r times the prove_index function and returns an empty
+    /// proof if no suitable candidate is found.
+    pub fn prove(setup: &Setup, set: &[Element]) -> Self {
+        for v in 0..setup.r {
```

Comment:
As mentioned in https://github.com/cardano-scaling/alba/pull/34#discussion_r1813060453, we should probably cut down `set` to `2 * n_p` elements to make sure the prover running time stays low.